Add pyproject.toml, poetry.lock, and Makefile for better project setup

jurassic-squid opened the pull request:
This pull request introduces three new files: `pyproject.toml`, `poetry.lock`, and `Makefile`. These files will streamline the setup process for contributors and ensure consistency across different environments.

The `pyproject.toml` file is a configuration file for Poetry, which is used to manage project dependencies. It specifies the project's dependencies and other metadata. The `poetry.lock` file is generated by Poetry and ensures that everyone uses the same versions of dependencies.

The `Makefile` contains several commands that automate common tasks:

`make install`: Installs the project's dependencies.
`make update`: Updates the project's dependencies, i.e. retrieves latest versions.
`make fmt`: Formats the code using autoflake, isort, and black.
`make lint`: Checks the code for errors using mypy, flake8, and bandit.
`make test`: Runs the tests.

To set up your environment as a contributor, you'll need to have Poetry installed. Then, you can run `make install` to install the project's dependencies. You can use the other make commands to format, lint, and test your code.

Please let me know if you have any questions or feedback.

jurassic-squid commented:
I can switch the dependencies if you'd rather go with [Ruff](https://docs.astral.sh/ruff/) for formatting and linting @mwermelinger? I haven't used it but it looks really good.

mwermelinger commented:
Thanks for this. I've not used Ruff, autoflake or bandit myself, so let's go with the tools you chose for the moment. We can always update to Ruff later, after we experiment with it and if it turns out to be better. I've assigned @densnow as reviewer as I'm on leave this week.

densnow commented on pyproject.toml:
```diff
@@ -0,0 +1,28 @@
+license = "BSD 3-Clause License"
+
+[tool.poetry.dependencies]
+python = "^3.10"
```
This is allowing versions of Python >=3.10.0 <4.0.0 anything up to a change in the major version number. When I ran the set up Python 3.11 was used in the environment . I believe something like `"~3.10"` (note the change from `^` to `~`) would only allow patch level changes hence keeping it to 3.10 see  [here](https://python-poetry.org/docs/dependency-specification/) for reference.

jurassic-squid replied:
Thanks for reviewing. I've made the Python version change you suggested. I had a quick look at the `pytype` [requirements](https://github.com/google/pytype#requirements) and it seems that 3.8 to 3.11 is supported, so perhaps there has been an update. I also removed `freezegun` from the tools because I don't think we will need that for this project. Let me know if you spot anything else.

densnow replied:
Thanks, Yes it looks like `pytype` is now supporting 3.11 as you say.  It might be worth raising an issue about this to discuss further. AFAIK `pytype` was the main reason for the Python 3.10 dependency in this project.

densnow commented:
Tested on Linux and everything seems to be working as expected now. This is a nice contribution to the project and gets  the ball rolling for a lot of other things. Many thanks @jurassic-squid 

In the future it would  be nice to add some documentation about the new workflow for contributors what do you think @mwermelinger.  For now though, this is ready to be merged.

densnow commented:
@mwermelinger I don't seem to have write permissions for this repo, only the M269 installer, so cannot squash and merge at present.

mwermelinger reviewed (changes requested):
Please update the `contributing.md` file in accordance with the use of `poetry` and the makefile so that the docs are always in sync w/ changes in code.

densnow commented:
@jurassic-squid  RE updated `contribution.md`: In general, I like the accessible  approach you have taken, and certainly agree about environment and version control being blockers. Please take the  points below as things which require further discussion and by no means as absolute truths.

- The [desktop](https://desktop.github.com/) app is another option for interfacing with GH which can abstract away some of the interaction with Git in the CLI. This could be worth mentioning as some students might be put off by a lot of terminal commands.
- Is a pure Windows environment as opposed to WSL an option for contributors? It comes down to the `pytype` dependency again and how it is not supported on Windows. Things like running tests with `-m` option would not work  and `poetry` might complain if `pytype` was absent
- If the point above is True, it follows that a WSL installation will have to be recommended for Windows users with links to documentation.
- It could be worth Mentioning to M269 students specifically that they can ask any questions re contributing on the VLE Tech forum as well as GH forums.
- Sometimes other projects mention that contributions do not necessarily have to be in the form of code. For example, correcting typos in documentation or updating rotten links etc. Is this something we want to mention?

jurassic-squid commented:
@densnow  Thanks for the feedback. I agree with all your points. I forgot that `pytype` doesn't work on Windows! I should probably add the step: 'run tests' before staging and commiting changes too. Always a good idea.

Could you point me towards the forums you mentioned? I'm not sure exactly which ones you mean.

I'll wait to hear back from Michel and then make changes.

mwermelinger commented on docs/contribution.md:
Thanks for the detailed instructions!

jurassic-squid replied:
Thanks for taking the time to look this over. I'll work on some revisions.

jurassic-squid commented on docs/contribution.md:
```diff
@@ -1,22 +1,154 @@
+
+This command tells Poetry to install the dependencies listed in the `pyproject.toml` file - a file that describes the project and the additional libraries it relies on.
+
+### 4. Activate the Environment
```
This was with Windows users in mind, as they wouldn't have been able to use `make`. But now that I'm reminded that Windows is not supported, it isn't needed!

jurassic-squid commented on docs/contribution.md:
```diff
@@ -1,22 +1,154 @@
\ No newline at end of file
+# Contributing to `allowed`
+
+Welcome to `allowed`! We appreciate your interest in contributing. Whether you're a seasoned developer or just getting started, we encourage and value all contributions.
```
Understood, will remove.

jurassic-squid commented on docs/contribution.md:
```diff
@@ -1,22 +1,154 @@
+
+### 5. Engage in Discussions
+After you've created a pull request, other developers might ask questions or provide feedback about your changes. Be sure to respond to these comments and address any feedback they provide. This collaborative process is a key part of open source development, and it helps to ensure that all changes are beneficial and free of bug
+
```
Oh, the irony...

jurassic-squid commented on docs/contribution.md:
```diff
@@ -1,22 +1,154 @@
+### 4. Create a Pull Request
+A "pull request" is a proposal to merge your changes into the main codebase. It allows other developers to review your changes and provide feedback before the changes are merged.
+
+To create a pull request, go to the GitHub page for your fork, switch to the branch that you just pushed, and click on the "New Pull Request" button.
```
I'll add instructions!

jurassic-squid commented:
I think I have addressed all the issues in `contributing.md`. I have ticked `Allow edits by maintainters`, so if there are any further changes you'd like to make, you can commit to the branch.  That might be easier. I won't be offended if you want to change/delete any bits!
